src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,2233 @@
+    step_mem->nfe++;
+    if (retval != 0)
+    {
+      arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                      MSG_ARK_RHSFUNC_FAILED, t);
+      return ARK_RHSFUNC_FAIL;
+    }
+
+    break;
+
+  default:
+    /* return with RHS failure if unknown mode is passed */
+    arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                    "Unknown full RHS mode");
+    return ARK_RHSFUNC_FAIL;
+  }
+
+  return ARK_SUCCESS;
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_TakeStepRKC:
+
+  This routine serves the primary purpose of the LSRKStepRKC module:
+  it performs a single RKC step (with embedding, if possible).
+
+  The output variable dsmPtr should contain estimate of the
+  weighted local error if an embedding is present; otherwise it
+  should be 0.
+
+  The input/output variable nflagPtr is generally used in ARKODE 
+  to gauge the convergence of any algebraic solvers. However, since 
+  the STS step routines do not involve an algebraic solve, this variable 
+  instead serves to identify possible ARK_RETRY_STEP returns within this 
+  routine.
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+                 ARK_RETRY_STEP indicates that the required stage
+                 number has reached the stage_max_limit with the
+                 current value of h. The step is then returned to
+                 adjust the step size.
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int lsrkStep_TakeStepRKC(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+{
+  int retval;
+  sunrealtype hmax, w0, w1, temp1, temp2, arg, bjm1, bjm2, mus, thjm1, thjm2,
+    zjm1, zjm2, dzjm1, dzjm2, d2zjm1, d2zjm2, zj, dzj, d2zj, bj, ajm1, mu, nu,
+    thj;
+  const sunrealtype onep54 = SUN_RCONST(1.54), c13 = SUN_RCONST(13.0),
+                    p8 = SUN_RCONST(0.8), p4 = SUN_RCONST(0.4);
+  ARKodeLSRKStepMem step_mem;
+
+  /* initialize algebraic solver convergence flag to success,
+     temporal error estimate to zero */
+  *nflagPtr = ARK_SUCCESS;
+  *dsmPtr   = ZERO;
+
+  /* access ARKodeLSRKStepMem structure */
+  retval = lsrkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return retval; }
+
+  sunrealtype* cvals = step_mem->cvals;
+  N_Vector* Xvecs    = step_mem->Xvecs;
+
+  /* Compute dominant eigenvalue and update stats */
+  if (step_mem->dom_eig_update)
+  {
+    retval = lsrkStep_ComputeNewDomEig(ark_mem, step_mem);
+    if (retval != ARK_SUCCESS) { return retval; }
+  }
+
+  sunrealtype ss =
+    SUNRceil(SUNRsqrt(onep54 * SUNRabs(ark_mem->h) * step_mem->spectral_radius));
+  ss = SUNMAX(ss, SUN_RCONST(2.0));
+
+  if (ss >= (sunrealtype)step_mem->stage_max_limit)
```

Comment:
This is closer to what was there before, but @Steven-Roberts was concerned about overflow from `SUNIceil` (since the maximum `sunindextype` is much smaller than the maximum `sunrealtype`), which is why the "ceiling" was converted to occur in floating-point arithmetic.

Reply:
Under `SUNIceil` the ceiling operation is still applied on the floating point type

Reply:
I see what you mean now. This is a safer comparison. It would be good to note this in a comment above the `ss` computation so we don't forget why it's this way in the future. Also, I think the `sunrealtype` cast can be removed (it's not used in the RKL step function).

Reply:
`sunrealtype` cast is removed.